dxf2nc.1
.\" -*- nroff -*-
.\" dxf2nc.1
.\" By: R.F. Smith <rsmith@xs4all.nl>
.\" $Date$
.\"
.Dd
.Os 
.Dt dxf2nc 1 CON
.Sh NAME
.Nm dxf2nc
.Nd Extract 
.Ql LINE 
and 
.Ql ARC 
entities from a Drawing eXchange Format 
.Pq Dq dxf
file and arranges them in a way suitable for optimized
cutting. Outputs a numeric code file for a Gerber cutter, with the lower left
corner of the bounding box of all the entities moved to 0,0.
.Sh SYNOPSIS
.Nm 
.Op Ar
.Sh DESCRIPTION
This manual page documents 
.Ql $Revision$
of the 
.Nm
program. 

The cutworks software that comes with a gerber cutter doesn't
automatically optimize the cutting paths it reads from dxf files. It
essentially cuts lines in the order it finds them in the dxf file.

This program reads a dxf file. The name of the file must end in 
.Ql .dxf
or 
.Ql .DXF , 
otherwise the program will report an error and quit. It
extracts all the
.Ql LINE 
and 
.Ql ARC
entities from it. It then searches through all these entities and assembles
connected entities into lists called contours. If necessary, the direction of
entities in a contour is changed so that all entities can be cut in one
continuous movement.

These contours and any remaining lines and arcs are then sorted in ascending
order from the left-bottom corner of their bounding box.

It also converts arcs into line segments. By default these line segments are
no more than 5 mm in length. It ignores the
.Dv $MEASUREMENT
variable in the dxf file because that is often not set correctly and
assumes that the units in the dxf file are millimeters.

The entities are then translated to Gerber numeric code, which is written
to the output file. The name of the new file is based on the
name of the old file, with the 
.Ql dxf 
extension replaced by 
.Ql nc .
So 
.Ql foo.dxf 
becomes 
.Ql foo.nc .

Gerber numeric code files are basically text files but do not contain
line breaks, which makes them hard to read. The 
.Xr ncfmt 1
utility can be used to display the file in a more human-readable format.

.Sh EXIT STATUS
.Ex -std
.\".Sh DIAGNOSTICS
.Sh COMPATIBILITY
The 
.Nm
program requires the Python interpreter. It was written for version 2.7, but
should be able to work with 3.x after fixing with
.Xr 2to3 .
.Sh SEE ALSO
.Xr python 1 ,
.Xr 2to3, 
.Xr ncfmt 1 .
.Rs
.%B Python Programming Language - Official Website
.%U http://www.python.org/
.Re
.Sh HISTORY
This program was written to stop having to avoid having to deal with cutworks.
.Sh AUTHOR
This manual and the 
.Nm
software were written by 
.An Roland Smith Aq rsmith@xs4all.nl .
.Pp
The latest version of this program is available at:
.Lk http://rsmith.home.xs4all.nl/software/ 
.Sh LICENSE
To the extent possible under law, Roland Smith has waived all copyright and
related or neighboring rights to this manual. This work is published from the
Netherlands. See 
.Lk http://creativecommons.org/publicdomain/zero/1.0/
.Pp
The
.Nm
program itself is released under the two-clause BSD license given below;
.Pp
Copyright \(co 2012 R.F. Smith <rsmith@xs4all.nl>. All rights reserved.
.Pp
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
.Bl -tag -width "foo" -compact
.It 1.
Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.
.It 2.
Redistributions in binary form must reproduce the above copyright notice, this
list of conditions and the following disclaimer in the documentation and/or
other materials provided with the distribution.
.El
.Pp
THIS SOFTWARE IS PROVIDED BY AUTHOR AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS
OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
EVENT SHALL AUTHOR OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\" EOF


dxf2nc.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Optimizes lines and arcs from a DXF file for cutting on a Gerber cutter,
# outputs G-codes for the Gerber cutter..
#
# Copyright © 2012 R.F. Smith <rsmith@xs4all.nl>. All rights reserved.
# $Date$
# 
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY AUTHOR AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL AUTHOR OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.

# System modules
import sys
import datetime
import os.path

# Local module.
import dxfgeom

__proginfo__ = ('dxf2nc [ver. ' + '$Revision$'[11:-2] + 
                '] ('+'$Date$'[7:-2]+')')

def nc_header(progname, bbox):
    '''Returns the start of the NC file.'''
    Li = (bbox[2]-bbox[0])/25.4
    Wi = (bbox[3]-bbox[1])/25.4
    s = "H1*M20*{}/L={:5.3f}/W={:5.3f}*N1*M15*".format(progname, Li, Wi)
    return s

def nc_footer():
    '''Returns the ending of the NC file.'''
    return 'M0*'

def newname(oldpath):
    '''Write the header of the NC file.'''
    oldname = os.path.basename(oldpath)
    if not oldname.endswith(('.dxf', '.DXF')):
        raise ValueError('not a DXF file!')
    oldbase = oldname[:-4]
    if len(oldbase) == 0:
        raise ValueError("zero-length file name!")
    rv = oldbase + '.nc'
    return rv

def main(argv):
    if len(argv) == 1:
        print __proginfo__
        print "Usage: {} [file.dxf ...]".format(argv[0])
        exit(1)
    del argv[0]
    for f in argv:
        try:
            outname = newname(f)
            ent = dxfgeom.read_entities(f)
        except ValueError:
            print "Cannot construct output filename. Skipping file '{}'.".format(f)
            print "A valid filename _must_ have a '.dxf' extension."
            print "And it must be more than just the extension."
            continue
        except IOError:
            print "Cannot open the file '{}'. Skipping it.".format(f)
            continue
        # Find entities
        lo = dxfgeom.find_entities("LINE", ent)
        lines = []
        if len(lo) > 0:
            lines = [dxfgeom.line_from_elist(ent, nn) for nn in lo]
        ao = dxfgeom.find_entities("ARC", ent)
        arcs = []
        if len(ao) > 0:
            arcs = [dxfgeom.arc_from_elist(ent, m) for m in ao]
        # TODO: Find bounding box.
        bbs = [ln.getbb() for ln in lines]
        bbs += [a.getbb() for a in arcs]
        bb = bbs[0]
        for add in bbs[1:]:
            bb = dxfgeom.merge_bb(bb, add)
        # TODO: Translate all elements so that bounding box starts at 0,0.
        dispx = -bb[0]
        dispy = -bb[1]
        for ln in lines:
            ln.move(dispx, dispy)
        for a in arcs:
            a.move(dispx, dispy)
        # Find contours
        (contours, remlines, remarcs) = dxfgeom.find_contours(lines, arcs)
        # Sort in y1, then in x1.
        contours.sort()
        remlines.sort()
        remarcs.sort()
        # Output
        outf = open(outname, 'w')
        outf.write(nc_header(outname[:-3], bb))
        for cn in contours:
            s1, s2 = cn.ncdata()
            outf.write(s1+s2)
        for l in remlines:
            s1, s2 = l.ncdata()
            outf.write(s1+s2)
        for a in remarcs:
            s1, s2 = a.ncdata()
            outf.write(s1+s2)
        outf.write(nc_footer())
        outf.close()

if __name__ == '__main__':
    main(sys.argv)
